packages/tools/src/tools/annotation/SplineROITool.ts

```diff
@@ -0,0 +1,895 @@
+      if (!viewport.getRenderingEngine()) {
+        console.warn('Rendering Engine has been destroyed');
+        return renderStatus;
+      }
+
+      let activeHandleCanvasCoords;
+
+      if (!isAnnotationVisible(annotationUID)) {
+        continue;
+      }
+
+      if (
+        !isAnnotationLocked(annotation) &&
+        !this.editData &&
+        activeHandleIndex !== null
+      ) {
+        // Not locked or creating and hovering over handle, so render handle.
+        activeHandleCanvasCoords = [canvasCoordinates[activeHandleIndex]];
+      }
+
+      if (activeHandleCanvasCoords || newAnnotation || highlighted) {
+        const handleGroupUID = '0';
+        drawHandlesSvg(
+          svgDrawingHelper,
+          annotationUID,
+          handleGroupUID,
+          canvasCoordinates,
+          {
+            color,
+            lineDash,
+            lineWidth,
+            handleRadius: '3',
+          }
+        );
+      }
+
+      if (splineConfig.showControlPointsConnectors) {
+        const controlPointsConnectors = [...canvasCoordinates];
+
+        // Connect the last point to the first one when the spline is closed
+        if (spline.closed) {
+          controlPointsConnectors.push(canvasCoordinates[0]);
+        }
+
+        drawPolylineSvg(
+          svgDrawingHelper,
+          annotationUID,
+          'controlPointsConnectors',
+          controlPointsConnectors,
+          {
+            color: 'rgba(255, 255,, 255, 0.5)',
+            lineDash,
+            lineWidth,
+          }
+        );
+      }
+
+      drawPolylineSvg(
+        svgDrawingHelper,
+        annotationUID,
+        'lineSegments',
+        splinePolyline,
+        {
+          color,
+          lineDash,
+          lineWidth,
+        }
+      );
+
+      renderStatus = true;
+    }
+
+    return renderStatus;
+  };
+
+  addControlPoint = (
+    evt: EventTypes.InteractionEventType,
+    annotation: SplineROIAnnotation
+  ) => {
+    const splineType = annotation.data.spline.type;
+    const splineConfig = this._getSplineConfig(splineType);
+    const maxDist = splineConfig.addControlPointLineDistance;
+
+    if (splineConfig.addControlPointEnabled === false) {
+      return;
+    }
+
+    const eventDetail = evt.detail;
+    const { element } = eventDetail;
+
+    const viewportIdsToRender = getViewportIdsWithToolToRender(
+      element,
+      this.getToolName()
+    );
+
+    const enabledElement = getEnabledElement(element);
+    const { renderingEngine, viewport } = enabledElement;
+    const { canvasToWorld } = viewport;
+
+    const spline = this._getAnnotationSpline(element, annotation);
+    const canvasPos = evt.detail.currentPoints.canvas;
+    const closestPointInfo = spline.getClosestPoint(canvasPos);
+
+    if (closestPointInfo.distance <= maxDist) {
+      spline.addControlPointAt(closestPointInfo.uValue);
+      annotation.data.handles.points = spline
+        .getControlPoints()
+        .map((p) => canvasToWorld(p));
```

Comment:
not sure why you are redoing the canvasToWorld here? shouldn't you only do it for the last added point?